utils/wasm-gen/src/generator/syscalls/invocator.rs

```diff
@@ -529,52 +515,139 @@ impl<'a, 'b> SysCallsInvocator<'a, 'b> {
-                .elements_mut();
-
-            // The end of insertion range is second-to-last index, as the last
-            // index is defined for `Instruction::End` of the function body.
-            // But if there's only one instruction in the function, then `0`
-            // index is used as an insertion point.
-            let last = if code.len() > 1 { code.len() - 2 } else { 0 };
-
-            let res = self.unstructured.int_in_range(0..=last).map(|pos| {
-                log::trace!(" -- Inserting into position {pos}");
-                code.splice(pos..pos, instructions.iter().cloned());
-            });
+        Ok(())
+    }
+
+    fn get_syscall_insertion_mapping<I>(
+        &mut self,
+        syscalls: I,
+        insert_into_funcs: &[usize],
+    ) -> Result<Vec<(usize, Vec<InvocableSysCall>)>>
+    where
+        I: Iterator<Item = InvocableSysCall>,
+    {
+        // Keep it sorted by func_id to perform binary search on it.
+        let mut insertion_mapping: Vec<(usize, Vec<InvocableSysCall>)> =
+            Vec::with_capacity(insert_into_funcs.len());
+        for syscall in syscalls {
+            let insert_into = *self.unstructured.choose(insert_into_funcs)?;
+            let partition =
+                insertion_mapping.partition_point(|&(func_id, _)| func_id < insert_into);
+
+            if !insertion_mapping.is_empty()
+                && partition != insertion_mapping.len()
+                && insertion_mapping[partition].0 == insert_into
+            {
+                insertion_mapping[partition].1.push(syscall);
+            } else {
+                insertion_mapping.insert(partition, (insert_into, vec![syscall]));
+            }
+        }
+        Ok(insertion_mapping)
+    }
+
+    fn insert_sys_calls_into_fn(
+        &mut self,
+        insert_into_fn: usize,
+        syscalls: Vec<InvocableSysCall>,
+    ) -> Result<()> {
+        log::trace!(
+            "Random data before inserting sys-calls invoke instructions into function {insert_into_fn} - {}",
+            self.unstructured.len()
+        );
+
+        let fn_code_len = self
+            .module
+            .code_section()
+            .expect("has at least one function by config")
+            .bodies()[insert_into_fn]
+            .code()
+            .elements()
+            .len();
+
+        // The end of insertion range is second-to-last index, as the last
+        // index is defined for `Instruction::End` of the function body.
+        // But if there's only one instruction in the function, then `0`
+        // index is used as an insertion point.
+        let last = if fn_code_len > 1 { fn_code_len - 2 } else { 0 };
+
+        let mut insertion_positions = syscalls
+            .into_iter()
+            .map(|syscall| Ok((self.unstructured.int_in_range(0..=last)?, syscall)))
+            .collect::<Result<Vec<_>>>()?;
+
+        // Sort in descending order.
+        insertion_positions.sort_by(|a, b| b.0.cmp(&a.0));
```

Comment:
please, give explanation above this that it gives guarantee that no inserting invocations instructions will intersect each other